### DISCUSSION TITLE
New dags are not getting listed in Airflow

### DISCUSSION THREAD

**Discussion by mandarbt** [apache/airflow#33194] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

Hello,

We are using airflow 2.4.6 version

We are creating dags in Dags folder using file operations ( We have template created for DAG and new DAG is created using this template with some changes given by user)
So whenever user creates a files it gets saved in DAGs folder - which is common for all 3 components - Airflow Webserver, Scheduler and Celery instances

Issue : When user creates a file - it takes minutes to show up in webserver, in fact we can not run that DAG using API as it is not available. 
We tried below settings in **_airflow.cfg_**

_dagbag_import_timeout = 120.0
dag_file_processor_timeout = 300
min_file_process_interval = 30

# Howewq often (in seconds) to scan the DAGs directory for new files. Default to 5 minutes.
dag_dir_list_interval = 30_

As mentioned above - we have set list interval as 30 seconds - and was expecting that after every 30 seconds , if there are new files then those will get processed and will be available in Webserver UI and for running the DAG as well

Can anyone help us out here as this is kind of critical blocker for us.

Thanks in Advance!
MB

### How to reproduce

We are creating dags in Dags folder using file operations ( We have template created for DAG and new DAG is created using this template with some changes given by user)
So whenever user creates a files it gets saved in DAGs folder - which is common for all 3 components - Airflow Webserver, Scheduler and Celery instances

Issue : When user creates a file - it takes minutes to show up in webserver, in fact we can not run that DAG using API as it is not available. Sometime it takes upto 30 mins to show up in UI

### Operating System

linux

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by pankajkoti** [comment]
hi, you have not specified how you have deployed Airflow. Can you check in the Airflow UI configuration tab what is the value for dag_dir_list_interval? Maybe something else(maybe some env var) is overriding this value.

Also, suggest you to upgrade your Airflow version to one of the latest versions to rule out the possibility of intermittent bugs.

Also, this looks more like a troubleshooting question and hence I am converting it to a discussion.

**Reply by mandarbt** [reply]
Hi Pankaj,

Thanks for quick reply - For Deployment query 
We have deployed airflow webserver, scheduler and celery on separate EC2 instances, DAG folder is shared across these instances. Each component ( Webserver, celery and scheduler has its own airflow.cfg)
Idea here is - we can scale out celery and scheduler instance if required.

One questions Pankaj -
Are above mentioned settings looks ok for our scenario or we are missing something there.
I will make sure these settings are same in both scheduler's airflow.cfg and webserver's airflow.cfg

Thanks!

**Reply by pankajkoti** [reply]
how is the DAG folder shared across the instances? what is the sync mechanism like?

**Reply by mandarbt** [reply]
We are using AWS EFS - so this EFS is mounted on each EC2 - like network drive

**Reply by mandarbt** [reply]
My observations are - when user creates first few dags using our endpoint ( which will internally does file operation and creates a new DAG file in the DAG folder) - then these dags get reflected within a minute but as number of dags increases then it takes time to reflect on webserver and as well as - user can not use airflow api to run those newly created DAGs

**Reply by pankajkoti** [reply]
https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/production-deployment.html

Suggest you taking a read of this document.

**Reply by pankajkoti** [reply]
I think DAGs are not getting synced somehow. You might need to debug and verify that DAGs are actually getting created in the DAGs folder and also should verify in the UI like I mentioned earlier what are the values of the config parameter. 

Did you make sure to restart your Airflow components scheduler, webserver, worker after changing the config parameter value to 30 seconds?

**Reply by mandarbt** [reply]
Hi Pankaj,

Yes, We are restarting the airflow components after changing airflow.cfg for each component.
We have observed that - when user creates a file - its there in DAG folder - but somehow its not getting processed by scheduler/airflow webserver and not getting synced with DB. It takes time, sometime it takes 20 mins to show up as DAG in airflow

**Reply by pankajkoti** [reply]
can you try to upgrade your Airflow setup to a new version and see if it still happens so?

**Reply by pankajkoti** [reply]
https://airflow.apache.org/docs/apache-airflow/stable/installation/upgrading.html

**Comment by potiuk** [comment]
> We have observed that - when user creates a file - its there in DAG folder - but somehow its not getting processed by scheduler/airflow webserver and not getting synced with DB. It takes time, sometime it takes 20 mins to show up as DAG in airflow

As @pankajkoti  explained - You need to explain the deployment (technologies used, way you run airflow). Depending on the deployment you have (especially syncong setting/configuration) - it might take your shared volume to distribute the files. Also then  - dependin on your scheduler settings it might take time for the scheduler (or dag file processor - depending on your deployment) to scan the files and to turn them into SerializedDag table entries. This is only when you will see the dags showing up. 

There are lots of parameters you can use to fine tune the behaviour of scheduler - depending on your particular deployment some parameters might be more important than others - I recommend you to read https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/scheduler.html#fine-tuning-your-scheduler-performance that explains all the important parameters , the knobs you can turn and ways changing some parameters might impact the behaviour.

In order to do that. I would recomend you an experiment (that only you can do) - add a DAG in the same way your users do - at the same time make sure to track what's going on in your scheduler and dag file processor logs. And login to your scheduler to see if the file is actually appearing in scheduler. You should see it when dag file processor will scan directories and find your new files. Correllating the events - i.e when you created, when they show up for scheduler, when they are parsed and when they are serialized to DB should give some answers where bottlenecks are. A lot depends on the way how you synchronize the dags. Common source of problems is that when you have a lot of dag files and your synchronization solution is  somewhat "custom" - for example it removes and then restores all the DAG files, or it changes mtime of all the DAG files all the time - rather than only those files that changed. Airflow scheduler by default priorities files that have recent mtime - (see https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#config-scheduler-file-parsing-sort-mode) so if this is changing continously, it might take quite a lot of time for Scheduler to finally scan the new files. You might want to check that. 

Gathering those logs is the first step to understand what's going on.

BTW. Your webserver does not need to see the DAG files  - they are read from the DB after DAG file processor scans them - it's even very wrong from the security perspective. Only scheduler/dag file processor and worker need to have the access in Airlfow 2. You should make sure your webserver does not have access to them.
